### PR TITLE
Add digital averaging

### PR DESCRIPTION
**Pull request by bencefr**
This PR introduces conversion of the 8-bit digital value to a 16-bit value stored in the buffer which for every bit of the original value has a double-bit with the following interpretation:
- 00: undefined
- 01: the channel was 0 during sampling
- 10: the channel was 1 during sampling
- 11: the channel had both 0 and 1 values

### CODE REVIEW

File: src/utils/bitConversion.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    (((b8 & 128) + 128) << 7) |
+    (((b8 & 64) + 64) << 6) |
+    (((b8 & 32) + 32) << 5) |
+    (((b8 & 16) + 16) << 4) |
+    (((b8 & 8) + 8) << 3) |
+    (((b8 & 4) + 4) << 2) |
+    (((b8 & 2) + 2) << 1) |
+    ((b8 & 1) + 1);
+
+export const doubleBitValue = (b16, n) => 3 & (b16 >> (n + n));
```
Comment (datenreisender) [inline src/utils/bitConversion.js:49]: I now have an idea what these two functions do but I probably will have a harder time in a year. So I suggest that you document them a bit and also write some tests for them.

You could probably use the examples from our chat about them and also what you wrote in the PR:
* 00: undefined
* 01: the channel was 0 during sampling
* 10: the channel was 1 during sampling
* 11: the channel had both 0 and 1 values
```js
bitTranslate(0b00000000).toString(2)
"0101010101010101"
bitTranslate(0b11111111).toString(2)
"1010101010101010"
(bitTranslate(0b00000000) | bitTranslate(0b11111111)).toString(2)
"1111111111111111"
```

File: src/components/Chart/Chart.jsx [hunk 2]
```diff
@@ -161,10 +162,7 @@ const Chart = ({ digitalChannelsEnabled = false }) => {
 
     const { data, bits } = options;
 
-    let numberOfBits =
-        windowDuration <= 3000000 && showDigitalChannels
-            ? nbDigitalChannels
-            : 0;
+    let numberOfBits = showDigitalChannels ? nbDigitalChannels : 0;
```
Comment (datenreisender) [inline src/components/Chart/Chart.jsx:165]: I assume you want to revert this change before taking this PR out of draft.

Comment (bencefr) [reply src/components/Chart/Chart.jsx]: I reverted this, but then I noticed that 3 sec limitation does not make sense in data logger. I suppose we would need a logging frequency based limitation.

**Comment by datenreisender** [comment]
Currently the representation of “bit was on and off” changes when zooming: When zooming in far enough, that state is represented by a point in the middle, like here on the left side and again on the right when going from "always off" to "always on": 
image

But when zooming out, “bit was on and off” is represented by points at the top and bottom with a line between them: 
image

I guess we will decide tomorrow which one we want, and then fix one of the two cases.

**Comment by datenreisender** [comment]
The digital channels change when saving and reloading a state. This should be the same data after capture and then after loading it again:
ScreenCapture after capture
ScreenCapture after load

File: src/components/Chart/data/dataSelector.js [hunk 3]
```diff
@@ -83,7 +84,7 @@ export default () => ({
             for (let i = 0; i < numberOfBits; ++i) {
                 const y = Number.isNaN(v)
                     ? undefined
-                    : (((bits[k] >> i) & 1) - 0.5) * 0.8;
+                    : [-0.5, -0.4, 0.4, 0][doubleBitValue(bits[k], i)];
```
Comment (nicolai-nordic) [inline src/components/Chart/data/dataSelector.js:87]: I can't really tell what is going on here :D Can we extract this to a function and give it a more descriptive name to help me and others like me? :) I don't know what these magic numbers (-0.5, -0.4 etc) refers to at all

Comment (bencefr) [reply src/components/Chart/data/dataSelector.js:87]: Best not to bother with that now, as I expect it will change with Marko's new double line and filled area representation. However the values are a the up and down state of the bit in [-0.5, 0.5] range with a slight margin, hence 0.4.

File: src/reducers/chartReducer.js [hunk 4]
```diff
@@ -124,28 +121,48 @@ export const chartWindowAction = (
...
         windowBegin: center - half,
         windowEnd: center + half,
-        windowDuration: duration,
+        windowDuration,
         yMin: y0,
         yMax: y1,
     };
 };
 
-export const chartReset = windowDuration =>
```
Comment (nicolai-nordic) [inline src/reducers/chartReducer.js:144]: Why remove this? It was quite nice imo, with its descriptive name,

Comment (bencefr) [reply src/reducers/chartReducer.js:144]: Because it wasn't used.

File: src/reducers/chartReducer.js [hunk 5]
```diff
@@ -88,17 +89,13 @@ export const chartCursorAction = (cursorBegin, cursorEnd) => ({
     cursorEnd,
 });
 
-export const chartWindowAction = (
+const chartWindowAction2 = (
```
Comment (nicolai-nordic) [inline src/reducers/chartReducer.js:92]: [minor] Not a huge fan of just appending 2 to the name. Perhaps the initial function should be changed to `prepareChartWindowAction` and this one be called `chartWindowAction` or something like that?

**Review by nicolai-nordic — APPROVED** [review]

File: src/reducers/chartReducer.js (hunk 5 above)
Comment (bencefr) [reply src/reducers/chartReducer.js:92]: I didn't want to change the name of the function everywhere where it was called :\
